File: Linked-Lists/19.js
// 19. Remove Nth Node From End of List
// Given the head of a linked list, remove the nth node from the end of the list and return its head.

var removeNthFromEnd = function (head, n) {
    if (!head.next) {
        return n > 1 ? head : null
    }
    let walker = head
    let length = 0
    while (walker) {
        length++
        walker = walker.next
    }
    let target = length - n
    walker = head
    if (target === 0) return walker.next
    let count = 0
    while (count < target - 1) {
        count++
        walker = walker.next
    }
    let nextNode = walker.next.next
    walker.next = nextNode
    return head
};

class ListNode {
    constructor(val, next) {
        this.val = (val === undefined ? 0 : val)
        this.next = (next === undefined ? null : next)
    }
    append(val) {
        const newNode = new ListNode(val)
        if (this.val === null) {
            this.val = newNode
            return
        }
        let walker = this
        while (walker.next) {
            walker = walker.next
        }
        walker.next = newNode
    }
    printAsArray() {
        let array = []
        let walker = this
        while (walker) {
            array.push(walker.val + ' ' + '->')
            walker = walker.next
        }
        array.push('null')
        console.log(array)
    }
}

const head = new ListNode(1)
head.append(2)
head.append(3)
head.append(4)
head.append(5)

let result = removeNthFromEnd(head, 2)
result.printAsArray()

File: Linked-Lists/19.py
# 19. Remove Nth Node From End of List
# Given the head of a linked list, remove the nth node from the end of the list and return its head.

class ListNode:
    def __init__(self, val = 0, next = None):
        self.val = val
        self.next = next

class Solution:
    def removeNthFromEnd(self, head: [ListNode], n: int) -> [ListNode]:
        walker = head
        length = 0
        while walker:
            walker = walker.next
            length += 1
        if length == n:
            return head.next
        trgt_idx = length - n
        idx = 1
        walker = head
        while idx != trgt_idx:
            walker = walker.next
            idx += 1
        removed_node = walker.next
        new_next = removed_node.next
        walker.next = new_next
        return head

node5 = ListNode(5)
node4 = ListNode(4, node5)
node3 = ListNode(3, node4)
node2 = ListNode(2, node3)
test_case = ListNode(1, node2)

s = Solution()
output = s.removeNthFromEnd(test_case, 2)

# Print new, modified list as a string
display = ""
walker = output
while walker:
    display += str(walker.val) + " -> "
    walker = walker.next
print(display + "None")